[Concept] Docs - Enumerations (#669)
exercism/cpp: concepts/enums/introduction.md

# Introduction

## Enumerations

So far we have mostly handled numbers and strings.

> To model the real world, we might want a limited number of options that a variable can take on.

Reviewer: (maybe) "options" -> "values"

> If you work for a skateboard manufacturer the deck material can be selected from maple, bamboo, or plastic.

Reviewer: can be shortened to: "In a skateboard factory the deck material ..."

> You could use integers to encode those values, but would have to use extra code to check if there is an invalid value coming from the system for the material.

Reviewer: I'm not sure if enums can really prevent invalid values since integers can be converted to enums without any kind of check.  
I'd rather argue that enums allow you to use expressive symbols (`maple`) instead of magic values (`0`), and they prevent mixing different types by accident.  
But that might be semantics.

Author: What do you think of this change?

Author: Okay, another version?

This is a case for `enumerations`.

## Example usage

The specific term for this kind of enumeration is `scoped enumeration`.
The deck material can be written like this.
Note the `enum class` keyword and the `;` at the end of the definition:

> 

Reviewer: "like this." is followed by a note instead of "this" example.  
Is there a better way to phrase/structure/order the two sentences and the example?

```cpp
enum class DeckMaterial {
    maple,
    bamboo,
    plastic
};
```

> Now, look at a pricing function in the skate shop and take note of the scope resolution operator (`::`) to specify an `enumerator` from the `enumeration`:

Reviewer: to specify -> that specifies (native English speaker, please correct me)

```cpp
double deck_price(double base_price, DeckMaterial material) {
    if(material == DeckMaterial::plastic) {
        return base_price * 0.9;
    }
    return base_price * 1.3;
}
```

## Scoping

Imagine you have a second `enumeration` for the material of the wheels:

```cpp
enum class WheelMaterial {
    steel,
    clay,
    plastic
};

bool comparison{DeckMaterial::plastic == WheelMaterial::plastic};

> // => false

Reviewer: false => does not compile

```
Although the wheels and the deck can both be made of _plastic_, the two cannot be confused.

> Each `enumeration` will have its `enumerators` in its own scope - its own namespace.

Reviewer: maybe put namespace in quotes

This is the reason they are called `scoped enumerations`.

~~~~exercism/advanced
## Unscoped Enumerations

You might think that there should be `unscoped` enumerations as well and you are correct.

> `Unscoped enumerations` lost favor to the version above, because they populate the global namespace.

Reviewer: is "to lose favor to sth." a common idiom? (@BethanyG)

Reviewer 2: Yeah - _technically_ I think `to lose favor to` works(ish) - as in there is nothing wrong grammar-wise.  

But as far as usage in idiom, it's:
-  `to lose favor with` (_a person or institution_), 
- `to lose favor as` (_a method, tool, strategy, or philosophy_), 
- `to lose favor among` (_a group_),   or 
- `to lose favor`  (_vague general approval or popularity_) .

You can say `lose favor compared to` - but that can sound a bit stuffy.  I think most AmEn speakers would use `as` or `with` in that circumstance instead.

For this reason you could not have two `unscoped enumerations` with the same `enumerators` like _plastic_ in the example above.

Also, `unscoped enumerations` implicitly convert to integers. 
Look at the example to see a surprising result:

```cpp
enum CitrusFruits {
    lemons,  // 0
    oranges, //1
};

enum IceCream {
    walnut, // 0

>     apples,      // 1

Reviewer: align the comments

};

> bool comparison{apples == oranges};

Reviewer: no criticism: I like that line!

// => true
```

> If you want to convert `scoped enumerations` into integers you can use `static_cast<int>`.

Reviewer: into -> to

## Example usage

The specific term for this kind of enumeration is `scoped enumeration`.
The snippet below shows how to write a `DeckMaterial` `enumeration`.
Note the `enum class` keyword and the `;` at the end of the definition:

```cpp
enum class DeckMaterial {
    maple,
    bamboo,
    plastic
};
```
Now, look at a pricing function in the skate shop and take note of the scope resolution operator (`::`) to specify an `enumerator` from the `enumeration`:

```cpp
double deck_price(double base_price, DeckMaterial material) {
    if(material == DeckMaterial::plastic) {
        return base_price * 0.9;
    }
    return base_price * 1.3;
}
```

## Scoping

Imagine you have a second `enumeration` for the material of the wheels:

```cpp
enum class WheelMaterial {
    steel,
    clay,
    plastic
};
```
Although the wheels and the deck can both be made of _plastic_, the two cannot be confused.
Each `enumeration` will have its `enumerators` in its own scope - its own `namespace`.
This is the reason they are called `scoped enumerations`.

~~~~exercism/advanced
## Unscoped Enumerations

You might think that there should be `unscoped` enumerations as well and you are correct.

> `Unscoped enumerations` lost favor to the version above, because they populate the global namespace.

Reviewer 2: ```suggestion
`Unscoped enumerations` have lost favor over time, because they can clutter up the global namespace.
```

Reviewer 2: Another word instead of `clutter` could be `crowd`.  And you could add a qualification like:
`Unscoped enumerations` have lost favor over time, because they can  unnecessarily crowd the global namespace.

Clutter can also be qualified as:

 "`Unscoped enumerations` have lost favor over time, because they can create clutter in the global namespace."

> For this reason you could not have two `unscoped enumerations` with the same `enumerators` like _plastic_ in the example above.

Reviewer 2: ```suggestion
Because of the shared namespace, you could not have two `unscoped enumerations` with the same `enumerators` like _plastic_ in the example above.
```

> `Unscoped enumerations` are becoming less popular, because they all share the same global namespace.

Reviewer 2: ```suggestion
`Unscoped enumerations` are becoming less popular because they all share the same global namespace.
```

Also, `unscoped enumerations` implicitly convert to integers. 

> Look at the example to see a surprising result:

Reviewer 2: ```suggestion
Look at the example below for a surprising result:
```
